File: search-api/src/main/scala/au/csiro/data61/magda/search/elasticsearch/ElasticSearchQueryer.scala
package au.csiro.data61.magda.search.elasticsearch

import scala.collection.JavaConverters._
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

import org.elasticsearch.search.aggregations.Aggregation
import org.elasticsearch.search.aggregations.bucket.filter.InternalFilter
import org.elasticsearch.search.sort.SortOrder

import com.sksamuel.elastic4s._
import com.sksamuel.elastic4s.ElasticDsl._
import com.sksamuel.elastic4s.searches.RichSearchResponse
import com.sksamuel.elastic4s.searches.SearchDefinition
import com.sksamuel.elastic4s.searches.aggs.AggregationDefinition
import com.sksamuel.elastic4s.searches.aggs.FilterAggregationDefinition
import com.sksamuel.elastic4s.searches.queries.QueryDefinition
import com.sksamuel.elastic4s.searches.queries.QueryStringQueryDefinition
import com.typesafe.config.Config

import akka.actor.ActorSystem
import akka.stream.Materializer
import au.csiro.data61.magda.api.FilterValue
import au.csiro.data61.magda.api.Query
import au.csiro.data61.magda.api.Specified
import au.csiro.data61.magda.api.Unspecified
import au.csiro.data61.magda.api.model.RegionSearchResult
import au.csiro.data61.magda.api.model.SearchResult
import au.csiro.data61.magda.model.misc._
import au.csiro.data61.magda.search.SearchStrategy._
import au.csiro.data61.magda.search.SearchQueryer
import au.csiro.data61.magda.search.SearchStrategy
import au.csiro.data61.magda.search.elasticsearch.ElasticSearchImplicits._
import au.csiro.data61.magda.search.elasticsearch.FacetDefinition.facetDefForType
import au.csiro.data61.magda.search.elasticsearch.Queries._
import au.csiro.data61.magda.util.ErrorHandling.RootCause
import au.csiro.data61.magda.util.SetExtractor
import org.elasticsearch.search.aggregations.support.AggregationPath.PathElement
import scala.collection.JavaConversions._
import org.elasticsearch.search.aggregations.InternalAggregation
import com.sksamuel.elastic4s.searches.queries.SimpleStringQueryDefinition
import com.sksamuel.elastic4s.analyzers.CustomAnalyzerDefinition

class ElasticSearchQueryer(indices: Indices = DefaultIndices)(
    implicit val config: Config,
    implicit val system: ActorSystem,
    implicit val ec: ExecutionContext,
    implicit val materializer: Materializer,
    implicit val clientProvider: ClientProvider) extends SearchQueryer {
  private val logger = system.log

  lazy val clientFuture: Future[TcpClient] = clientProvider.getClient(system.scheduler, logger, ec)

  val ESCAPE_REGEX = "([\\+\\-=!\\(\\)\\{\\}\\[\\]\\^\"~\\?:/\\\\]|&&|\\|\\|)".r
  val REMOVE_REGEX = "(?i)((^|\\s)(AND|OR)(\\s|$)|[<>])".r
  val ESCAPE_AGG_NAME_REGEX = "[\\[\\]\\.]".r

  val DATASETS_LANGUAGE_FIELDS = Seq("title", "description", "publisher.name", "distributions.title", "distributions.description", "keyword", "theme")
  val ASPECT_LANGUAGE_FIELDS = Seq("value")

  override def search(inputQuery: Query, start: Long, limit: Int, requestedFacetSize: Int) = {
    val inputRegionsList = inputQuery.regions.toList

    clientFuture.flatMap { implicit client =>
      val fullRegionsFutures = inputRegionsList.map(resolveFullRegion)
      val fullRegionsFuture = Future.sequence(fullRegionsFutures)

      Future.sequence(Seq(fullRegionsFuture, client.execute(buildQueryWithAggregations(inputQuery, start, limit, MatchAll, requestedFacetSize)).flatMap(response =>
        if (response.totalHits > 0)
          Future.successful((response, MatchAll))
        else
          client.execute(buildQueryWithAggregations(inputQuery, start, limit, MatchPart, requestedFacetSize)).map((_, MatchPart)))
      ))
    } map {
      case Seq(fullRegions: List[Option[Region]], (response: RichSearchResponse, strategy: SearchStrategy)) =>
        val newQueryRegions =
          inputRegionsList
            .zip(fullRegions)
            .filter(_._2.isDefined)
            .map(_._2.get)
            .map(Specified.apply).toSet ++ inputQuery.regions.filter(_.isEmpty)

        val outputQuery = inputQuery.copy(regions = newQueryRegions)

        buildSearchResult(outputQuery, response, strategy, requestedFacetSize)
    } recover {
      case RootCause(illegalArgument: IllegalArgumentException) =>
        logger.error(illegalArgument, "Exception when searching")
        failureSearchResult(inputQuery, "Bad argument: " + illegalArgument.getMessage)
      case e: Throwable =>
        logger.error(e, "Exception when searching")
        failureSearchResult(inputQuery, "Unknown error")
    }
  }

  /**
   * Calculate the ideal facet size for this query - we have to take the facet exact match queries into account because
   * if we always passed through the requested facetSize, we might end up in a situation where requested facet size
   * was 1 but the query had 3 exact match queries for publishers in it - we'd only know the hitCount for the first
   * exact match because we only asked for 1 facet from ES.
   */
  private def idealFacetSize(facetType: FacetType, query: Query, requestedFacetSize: Int) = {
    val queryFacetSize = FacetDefinition.facetDefForType(facetType).exactMatchQueries(query).size

    Math.max(queryFacetSize, requestedFacetSize)
  }

  /**
   * Turns an ES response into a magda SearchResult.
   */
  def buildSearchResult(query: Query, response: RichSearchResponse, strategy: SearchStrategy, facetSize: Int): SearchResult = {
    val aggsMap = response.aggregations.map
    new SearchResult(
      strategy = Some(strategy),
      query = query,
      hitCount = response.getHits.totalHits().toInt,
      dataSets = response.to[DataSet].map(_.copy(years = None)).toList,
      facets = Some(FacetType.all.map { facetType =>
        val definition = facetDefForType(facetType)

        new Facet(
          id = facetType.id,
          options = {
            // Filtered options are the ones that partly match the user's input... e.g. "Ballarat Council" for input "Ballarat"
            val filteredOptions =
              (aggsMap.get(facetType.id + "-filter") match {
                case Some(filterAgg) => definition.extractFacetOptions(filterAgg.getProperty(facetType.id).asInstanceOf[Aggregation])
                case None            => Nil
              }).filter(definition.isFilterOptionRelevant(query))
                .map(_.copy(matched = true))

            // filteredExact aggregations are those that exactly match a filter (e.g. "Ballarat Council" exactly) but are also filtered by
            // the rest of the query - we use this to filter the exact options below and make sure we don't show 0 results for a filtered
            // aggregation that does actually have results.
            val filteredExact = definition.exactMatchQueries(query)
              .map {
                case (name, query) => (
                  name,
                  aggsMap(facetType.id + "-exact-" + name + "-filter")
                )
              }
              .map {
                case (name, agg: InternalFilter) => name -> agg.getDocCount
              }
              .toMap

            // Exact options are for when a user types a correct facet name exactly but we have no hits for it, so we still want to
            // display it to them to show them that it does *exist* but not for this query
            val exactOptions =
              definition.exactMatchQueries(query)
                .map {
                  case (name, query) =>
                    val value = facetType.id + "-exact-" + name
                    val aggProp = List(value).asJava
                    (
                      name,
                      aggsMap(facetType.id + "-global").asInstanceOf[InternalAggregation].getProperty(aggProp)
                    )
                }
                .flatMap {
                  case (name, agg: InternalFilter) =>
                    if (agg.getDocCount > 0 && filteredExact.get(name).getOrElse(0l) == 0l) {
                      Some(FacetOption(name.getOrElse(config.getString("strings.unspecifiedWord")), 0, matched = true))
                    } else None
                }
                .toSeq

            val alternativeOptions =
              definition.extractFacetOptions(
                aggsMap(facetType.id + "-global")
                  .getProperty("filter").asInstanceOf[Aggregation]
                  .getProperty(facetType.id).asInstanceOf[Aggregation]
              )

            definition.truncateFacets(query, filteredOptions, exactOptions, alternativeOptions, facetSize)
          }
        )
      }.toSeq)
    )
  }

  /** Converts from a general search strategy to the actual elastic4s method that will combine a number of queries using that strategy.*/
  implicit def strategyToCombiner(strat: SearchStrategy): Iterable[QueryDefinition] => QueryDefinition = strat match {
    case MatchAll  => must
    case MatchPart => should
  }

  /** Builds an elastic search query out of the passed general magda Query */
  def buildQuery(query: Query, start: Long, limit: Int, strategy: SearchStrategy) = {
    ElasticDsl.search(indices.getIndex(config, Indices.DataSetsIndex) / indices.getType(Indices.DataSetsIndexType))
      .limit(limit)
      .start(start.toInt)
      .query(queryToQueryDef(query, strategy))
  }

  /** Same as {@link #buildQuery} but also adds aggregations */
  def buildQueryWithAggregations(query: Query, start: Long, limit: Int, strategy: SearchStrategy, facetSize: Int) = addAggregations(buildQuery(query, start, limit, strategy), query, strategy, facetSize)

  /** Builds an empty dummy searchresult that conveys some kind of error message to the user. */
  def failureSearchResult(query: Query, message: String) = new SearchResult(
    query = query,
    hitCount = 0,
    dataSets = Nil,
    errorMessage = Some(message)
  )

  /** Adds standard aggregations to an elasticsearch query */
  def addAggregations(searchDef: SearchDefinition, query: Query, strategy: SearchStrategy, facetSize: Int) = {
    val aggregations: List[AggregationDefinition] =
      FacetType.all.flatMap(facetType =>
        aggsForFacetType(query, facetType, strategy, idealFacetSize(facetType, query, facetSize))).toList

    searchDef.aggregations(aggregations)
  }

  /** Gets all applicable ES aggregations for the passed FacetType, given a Query */
  def aggsForFacetType(query: Query, facetType: FacetType, strategy: SearchStrategy, facetSize: Int): List[AggregationDefinition] = {
    val facetDef = facetDefForType(facetType)

    // Sub-aggregations of "global" aggregate on all datasets independently of the query passed in.
    val globalAgg =
      aggregation
        .global(facetType.id + "-global")
        .aggs(alternativesAggregation(query, facetDef, strategy, facetSize) :: exactMatchAggregations(query, facetType, facetDef, strategy))
        .asInstanceOf[AggregationDefinition]

    val partialMatchesAggs =
      if (facetDef.isRelevantToQuery(query))
        exactMatchAggregations(query, facetType, facetDef, strategy, "-filter") :+
          // If there's details in the query that relate to this facet
          // then create an aggregation that shows all results for this facet that partially match the details
          // in the query... this is useful if say the user types in "Ballarat", we can suggest "Ballarat Council"
          aggregation
          .filter(facetType.id + "-filter")
          .filter(facetDef.filterAggregationQuery(query))
          .aggs(facetDef.aggregationDefinition(facetSize))
          .asInstanceOf[AggregationDefinition]
      else
        List()

    partialMatchesAggs :+ globalAgg
  }

  /**
   * The exact match aggs are for situations where the user puts in a free text facet - we want
   * to see whether that exists in the system at all even if it has no hits with their current
   * query, in order to more helpfully correct their search if they mispelled etc.
   */
  def exactMatchAggregations(query: Query, facetType: FacetType, facetDef: FacetDefinition, strategy: SearchStrategy, suffix: String = ""): List[FilterAggregationDefinition] =
    facetDef.exactMatchQueries(query).map {
      case (name, query) =>
        aggregation.filter(facetType.id + "-exact-" + name + suffix).filter(query)
    }.toList

  /**
   * The alternatives aggregation shows what other choices are available if the user wasn't
   * filtering on this facet - e.g. if I was searching for datasets from a certain publisher,
   * this shows me other publishers I could search on instead
   */
  def alternativesAggregation(query: Query, facetDef: FacetDefinition, strategy: SearchStrategy, facetSize: Int) =
    aggregation
      .filter("filter")
      .filter(queryToQueryDef(facetDef.removeFromQuery(query), strategy))
      .aggs(facetDef.aggregationDefinition(facetSize))

  /**
   * Accepts a seq - if the seq is not empty, runs the passed fn over it and returns the result as Some, otherwise returns None.
   */
  private def setToOption[X, Y](seq: Set[X])(fn: Set[X] => Y): Option[Y] = seq match {
    case SetExtractor() => None
    case x              => Some(fn(x))
  }

  /** Processes a general magda Query into a specific ES QueryDefinition */
  private def queryToQueryDef(query: Query, strategy: SearchStrategy): QueryDefinition = {
    val processedQuote = query.quotes.map(cleanStringForEs).map(quote => s""""${quote}"""") match {
      case SetExtractor() => None
      case xs             => Some(xs.reduce(_ + " " + _))
    }

    val stringQuery: Option[String] = {
      (query.freeText.map(cleanStringForEs), processedQuote) match {
        case (None, None)                   => None
        case (None, some)                   => some
        case (some, None)                   => some
        case (Some(freeText), Some(quotes)) => Some(freeText + " " + quotes)
      }
    }

    val operator = strategy match {
      case MatchAll  => "and"
      case MatchPart => "or"
    }

    val clauses: Seq[Traversable[QueryDefinition]] = Seq(
      stringQuery.map { innerQuery =>
        val queryDef = new SimpleStringQueryDefinition(innerQuery)
          .defaultOperator(operator)
          .quoteFieldSuffix(".quote")
          .analyzeWildcard(true)

        ("_all" +: DATASETS_LANGUAGE_FIELDS).foldRight(queryDef) { (field, queryDef) =>
          queryDef.field(field)
        }
      },
      setToOption(query.publishers)(seq => should(seq.map(publisherQuery(strategy)))),
      setToOption(query.formats)(seq => should(seq.map(formatQuery))),
      dateQueries(query.dateFrom, query.dateTo),
      setToOption(query.regions)(seq => should(seq.map(region => regionIdQuery(region, indices))))
    )

    strategy(clauses.flatten)
  }

  def cleanStringForEs(string: String): String = {
    val cleaned = REMOVE_REGEX.replaceAllIn(ESCAPE_REGEX.replaceAllIn(string, charMatch => s"\\\\\\${charMatch.matched}"), " ")

    if (cleaned.replace("\\", "").equals(string.replace("\\", ""))) {
      cleaned
    } else {
      cleanStringForEs(cleaned)
    }
  }

  override def searchFacets(facetType: FacetType, facetQuery: String, generalQuery: Query, start: Long, limit: Int): Future[FacetSearchResult] = {
    val facetDef = facetDefForType(facetType)

    clientFuture.flatMap { client =>
      // First do a normal query search on the type we created for values in this facet
      client.execute(ElasticDsl.search in indices.getIndex(config, Indices.DataSetsIndex) / indices.getType(indices.typeForFacet(facetType)) query facetQuery start start.toInt limit limit)
        .flatMap { response =>
          response.totalHits match {
            case 0 => Future(FacetSearchResult(0, Nil)) // If there's no hits, no need to do anything more
            case _ =>
              val hitNames: Seq[String] = response.getHits.asScala.map(hit => hit.getSource.get("value").toString).toSeq

              // Create a dataset filter aggregation for each hit in the initial query
              val filters = hitNames.map(name =>
                aggregation.filter(name).filter(facetDef.exactMatchQuery(Specified(name))))

              // Do a datasets query WITHOUT filtering for this facet and  with an aggregation for each of the hits we
              // got back on our keyword - this allows us to get an accurate count of dataset hits for each result
              client.execute {
                buildQuery(facetDef.removeFromQuery(generalQuery), 0, 0, MatchAll).aggs(filters)
              } map { aggQueryResult =>
                val aggregations = aggQueryResult.aggregations.map
                  .mapValues {
                    case bucket: InternalFilter => new FacetOption(
                      value = bucket.getName,
                      hitCount = bucket.getDocCount
                    )
                  }

                FacetSearchResult(
                  hitCount = response.totalHits,
                  options = hitNames.map { hitName => aggregations(hitName) }
                )
              }
          }
        }
    }
  }

  override def searchRegions(query: String, start: Long, limit: Int): Future[RegionSearchResult] = {
    clientFuture.flatMap { client =>
      client.execute(
        ElasticDsl.search(indices.getIndex(config, Indices.RegionsIndex) / indices.getType(Indices.RegionsIndexType))
          query { matchPhrasePrefixQuery("regionName", query) }
          start start.toInt
          limit limit
          sortBy (
            fieldSort("order") order SortOrder.ASC,
            scoreSort order SortOrder.DESC
          )
            sourceExclude "geometry"
      ).flatMap { response =>
          response.totalHits match {
            case 0 => Future(RegionSearchResult(query, 0, List())) // If there's no hits, no need to do anything more
            case _ => Future(RegionSearchResult(query, response.totalHits, response.to[Region].toList))
          }
        }
    }
  }

  def resolveFullRegion(queryRegionFV: FilterValue[Region])(implicit client: TcpClient): Future[Option[Region]] = {
    queryRegionFV match {
      case Specified(region) =>
        client.execute(ElasticDsl.search(indices.getIndex(config, Indices.RegionsIndex) / indices.getType(Indices.RegionsIndexType))
          query { idsQuery((region.queryRegion.regionType + "/" + region.queryRegion.regionId).toLowerCase) } start 0 limit 1 sourceExclude "geometry")
          .flatMap { response =>
            response.totalHits match {
              case 0 => Future(None)
              case _ => Future(response.to[Region].headOption)
            }
          }
      case Unspecified() => Future(None)
    }

  }
}

object ElasticSearchQueryer {
  def apply(implicit config: Config, system: ActorSystem, ec: ExecutionContext, materializer: Materializer, clientProvider: ClientProvider) = new ElasticSearchQueryer()
}

File: magda-search-api-node/src/search/ElasticSearchQueryer.ts
import { Client, ApiResponse } from "@elastic/elasticsearch";
import _ = require("lodash");

import {
    Query,
    FacetType,
    FacetSearchResult,
    Region,
    FacetOption
} from "../model";
import SearchQueryer from "./SearchQueryer";
import getFacetDefinition from "./getFacetDefinition";

const client = new Client({ node: "http://localhost:9200" });

type LanguageField = {
    path: string;
    boost?: number;
};

const DATASETS_LANGUAGE_FIELDS: LanguageField[] = [
    { path: "title", boost: 50 },
    { path: "description", boost: 2 },
    { path: "publisher.name" },
    { path: "keywords", boost: 10 },
    { path: "themes" }
];
const NON_LANGUAGE_FIELDS: LanguageField[] = [
    { path: "_id" },
    { path: "catalog" },
    { path: "accrualPeriodicity" },
    { path: "contactPoint.identifier" },
    { path: "publisher.acronym" }
];

export default class ElasticSearchQueryer implements SearchQueryer {
    constructor(
        readonly datasetsIndexVersion: number,
        readonly regionsIndexVersion: number
    ) {}

    protected getDatasetsIndex(): string {
        return "datasets" + this.datasetsIndexVersion;
    }

    protected getRegionsIndex(): string {
        return "regions" + this.regionsIndexVersion;
    }

    async searchFacets(
        facetType: FacetType,
        generalQuery: Query,
        start: number,
        limit: number,
        facetQuery: string | undefined
    ): Promise<FacetSearchResult> {
        const facetDef = getFacetDefinition(facetType);

        const esQueryBody = {
            index: "publishers4",
            body: {
                query: {
                    dis_max: {
                        tie_breaker: 0,
                        queries: [
                            {
                                match_phrase_prefix: {
                                    value: facetQuery || ""
                                }
                            },
                            {
                                match_phrase_prefix: {
                                    acronym: facetQuery || ""
                                }
                            }
                        ]
                    }
                }
            },
            size: limit,
            from: start
        };
        // console.log(JSON.stringify(esQueryBody, null, 2));
        const result: ApiResponse = await client.search(esQueryBody);

        const { body } = result;

        if (body.totalHits === 0) {
            return { hitCount: 0, options: [] };
        } else {
            type Hit = {
                value: string;
                identifier?: string;
            };

            const hits: Hit[] = body.hits.hits.map((hit: any) => hit._source);

            // Create a dataset filter aggregation for each hit in the initial query
            const filters = hits.reduce(
                (soFar: any, { value }: Hit) => ({
                    ...soFar,
                    [value]: {
                        filter: facetDef.exactMatchQuery(value)
                    }
                }),
                {}
            );

            // Do a datasets query WITHOUT filtering for this facet and  with an aggregation for each of the hits we
            // got back on our keyword - this allows us to get an accurate count of dataset hits for each result

            const generalEsQueryBody = {
                from: 0,
                size: 0,
                body: {
                    query: await this.buildESDatasetsQuery(
                        facetDef.removeFromQuery(generalQuery)
                    ),
                    aggs: filters
                },
                index: this.getDatasetsIndex()
            };
            // console.log(JSON.stringify(generalEsQueryBody, null, 2));
            const resultWithout = await client.search(generalEsQueryBody);

            console.log(JSON.stringify(resultWithout.body, null, 2));

            const aggregationsResult = _(resultWithout.body.aggregations as {
                [aggName: string]: any;
            })
                .map((value, key) => ({
                    countErrorUpperBound: 0,
                    hitCount: value.doc_count || 0,
                    matched: false,
                    value: key
                }))
                .keyBy("value")
                .value();

            const options: FacetOption[] = _(hits)
                .map(hit => ({
                    ...aggregationsResult[hit.value],
                    identifier: hit.identifier
                }))
                .sortBy(hit => -hit.hitCount)
                .drop(start)
                .take(limit)
                .value();

            return {
                hitCount: resultWithout.body.hits.total,
                options
            };
        }
    }

    //   def augmentWithBoostRegions(query: Query)(implicit client: ElasticClient): Future[Query] = {
    //     val regionsFuture = query.freeText.filter(_.length > 0).map(freeText => client.execute(
    //       ElasticDsl.search(indices.getIndex(config, Indices.RegionsIndex))
    //         query(matchQuery("regionSearchId", freeText).operator("or"))
    //         limit 50
    //       ).map {
    //         case ESGenericException(e) => throw e
    //         case results: RequestSuccess[SearchResponse] => {
    //           results.result.totalHits match {
    //             case 0 => Set[Region]() // If there's no hits, no need to do anything more
    //             case _ => results.result.to[Region].toSet
    //           }
    //         }
    //       }
    //     ).getOrElse(Future(Set[Region]()))
    //     regionsFuture.map(regions => query.copy(boostRegions = regions))
    //   }

    async getBoostRegions(query: Query): Promise<Region[]> {
        if (!query.freeText || query.freeText === "") {
            return [];
        }

        const regionsResult = await client.search({
            index: this.getRegionsIndex(),
            body: {
                query: {
                    match: {
                        regionSearchId: {
                            query: query.freeText,
                            operator: "or"
                        }
                    }
                }
            },
            size: 50
        });

        return regionsResult.body.hits.hits.map((x: any) => x._source);
    }

    async buildESDatasetsQuery(query: Query): Promise<any> {
        const boostRegions = await this.getBoostRegions(query);

        const geomScorerQueries = boostRegions.map(
            this.regionToGeoshapeQuery,
            this
        );

        const qualityFactor = {
            filter: {
                term: {
                    hasQuality: true
                }
            },
            field_value_factor: {
                field: "quality",
                missing: 1
            }
        };

        const esQuery = {
            function_score: {
                query: this.queryToEsQuery(query, boostRegions),
                functions: [
                    {
                        weight: 1
                    },
                    qualityFactor,
                    {
                        filter: {
                            bool: {
                                should: geomScorerQueries
                            }
                        },
                        weight: 1
                    }
                ],
                score_mode: "sum"
            }
        };

        return esQuery;
    }

    queryToEsQuery(query: Query, boostRegions: Region[]): any {
        const freeText = (() => {
            const sanitisedFreeText =
                !query.freeText || query.freeText === "" ? "*" : query.freeText;
            const textQuery = this.textQuery(sanitisedFreeText);

            if (boostRegions.length === 0) {
                return textQuery;
            } else {
                const regionNames = _(boostRegions)
                    .flatMap(region => [
                        region.regionName,
                        region.regionShortName
                    ])
                    .filter(x => x && x !== "")
                    .sortBy(a => a.length)
                    .value();

                // Remove these regions from the text
                const textWithoutRegions = regionNames.reduce(
                    (soFar, currentRegion) =>
                        soFar.split(currentRegion).join(""),
                    sanitisedFreeText
                );

                const textQueryNoRegions = this.textQuery(
                    textWithoutRegions.length > 0 ? textWithoutRegions : "*"
                );
                const geomScorerQueries = boostRegions.map(
                    this.regionToGeoshapeQuery,
                    this
                );

                return {
                    bool: {
                        should: [
                            textQuery,
                            {
                                bool: {
                                    must: [
                                        textQueryNoRegions,
                                        ...geomScorerQueries
                                    ]
                                }
                            }
                        ],
                        minimum_should_match: 1
                    }
                };
            }
        })();

        return {
            bool: {
                must: [
                    freeText
                    //TODO: Put other queries in here
                ]
            }
        };
    }

    textQuery(inputText: string) {
        const simpleQueryStringQuery = {
            query: inputText,
            default_operator: "and",
            quote_field_suffix: ".quote"
        };

        // Surprise! english analysis doesn't work on nested objects unless you have a nested query, even though
        // other analysis does. So we do this silliness
        const distributionsEnglishQuery = {
            nested: {
                path: "distributions",
                score_mode: "max",
                query: {
                    simple_query_string: {
                        query: inputText,
                        fields: [
                            "distributions.title",
                            "distributions.description",
                            "distributions.format"
                        ],
                        default_operator: "and"
                    }
                }
            }
        };

        /**
         * Unfortunately, when default operator is AND, we can't put NON_LANGUAGE_FIELDS & DATASETS_LANGUAGE_FIELDS
         * into one SimpleStringQuery as they have different searchAnalylzer
         * It will result a term like +(catalog:at | _id:at)  will never be matched
         * We need to fix on our side as elasticsearch won't know our intention for this case
         */
        const queries = [
            {
                bool: {
                    should: [
                        {
                            simple_query_string: {
                                ...simpleQueryStringQuery,
                                fields: DATASETS_LANGUAGE_FIELDS.map(
                                    fieldDefToEs
                                )
                            }
                        },
                        {
                            simple_query_string: {
                                ...simpleQueryStringQuery,
                                fields: NON_LANGUAGE_FIELDS.map(fieldDefToEs)
                            }
                        }
                    ],
                    minimum_should_match: 1
                }
            },
            distributionsEnglishQuery
        ];

        return {
            dis_max: {
                tie_breaker: 0.2,
                queries
            }
        };
    }

    regionToGeoshapeQuery(region: Region) {
        return {
            geo_shape: {
                "spatial.geoJson": {
                    indexed_shape: {
                        index: this.getRegionsIndex(),
                        type: "regions",
                        id: region.regionSearchId,
                        path: "geometry"
                    }
                }
            }
        };
    }
}

function fieldDefToEs(def: LanguageField): string {
    if (typeof def.boost !== "undefined") {
        return `${def.path}^${def.boost}`;
    } else {
        return def.path;
    }
}
